File: 120.三角形最小路径和.js
/*
 * @lc app=leetcode.cn id=120 lang=javascript
 *
 * [120] 三角形最小路径和
 */

// @lc code=start
/**
 * @param {number[][]} triangle
 * @return {number}
 */
var minimumTotal = function (triangle) {
    if (!triangle.length) {
        return 0;
    }

    const readRow = (index, step, min) => {
        if (!step.length) {
            return readRow(index + 1, triangle[index], triangle[index][0]);
        }
        if (!triangle[index]) {
            return min;
        }

        let nextStep = [];
        let minStep = Infinity;
        for (let i = 0; i < triangle[index].length; i++) {
            const ele = triangle[index][i];

            let leftChild = step[i - 1] !== undefined ? step[i - 1] : Infinity;
            let rightChild = step[i] !== undefined ? step[i] : Infinity;
            nextStep[i] = Math.min(leftChild, rightChild) + ele;
            if (nextStep[i] < minStep) {
                minStep = nextStep[i]
            }
        }
        return readRow(index + 1, nextStep, minStep);
    }

    return readRow(0, [], 0);
};
// @lc code=end


let res = minimumTotal([
    [1], [-5, -2], [3, 6, 1], [-1, 2, 4, -3]
]);
console.log(res);

File: 120.三角形最小路径和.py
#
# @lc app=leetcode.cn id=120 lang=python3
#
# [120] 三角形最小路径和
#

from typing import List
# @lc code=start


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle)
        if n < 1:
            return 0

        dp = [0] * n
        for i in range(n-1, -1, -1):
            m = len(triangle[i])
            if i == n - 1:
                dp[i] = triangle[i]
            else:
                dp[i] = [0] * m
                for j in range(0, m):
                    dp[i][j] = triangle[i][j] + min(dp[i+1][j], dp[i+1][j + 1])
        return dp[0][0]


# @lc code=end

solution = Solution()
ans = solution.minimumTotal([
    [2],
    [3, 4],
    [6, 5, 7],
    [4, 1, 8, 3]
    # [1]
])
print(ans)
